### DISCUSSION TITLE
How to render JSON with one HTTP header, and render an HTML page with another header, in Next.js?

### DISCUSSION THREAD

**Discussion by lancejpollard** [vercel/next.js#48221] [Help] [answered]
### Summary

I saw a cool thing on I think the Deno.js website, where if you visit https://deno.land/std@0.182.0/fmt/colors.ts in the browser, it shows you the rendered thing:

<img width="1226" alt="Screenshot 2023-04-10 at 4 19 05 PM" src="https://user-images.githubusercontent.com/28718/231017150-b55807f6-e19a-43e7-b42c-b992481359ae.png">

But if you `curl` it, it gives you another thing:

```
$ curl https://deno.land/std@0.182.0/fmt/colors.ts
// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
// This module is browser compatible.
// A module to print ANSI terminal colors. Inspired by chalk, kleur, and colors
// on npm.

/**
 * String formatters and utilities for dealing with ANSI color codes.
 *
 * This module is browser compatible.
 *
 * This module supports `NO_COLOR` environmental variable disabling any coloring
 * if `NO_COLOR` is set.
 *
 * @example
 * ```typescript
 * import {
 *   bgBlue,
 *   bgRgb24,
 *   bgRgb8,
 *   bold,
 *   italic,
 *   red,
 *   rgb24,
 *   rgb8,
 * } from "https://deno.land/std@$STD_VERSION/fmt/colors.ts";
 *
 * console.log(bgBlue(italic(red(bold("Hello, World!")))));
 *
 * // also supports 8bit colors
 *
 * console.log(rgb8("Hello, World!", 42));
 *
 * console.log(bgRgb8("Hello, World!", 42));
 *
 * // and 24bit rgb
 *
 * console.log(rgb24("Hello, World!", {
 *   r: 41,
 *   g: 42,
 *   b: 43,
 * }));
 *
 * console.log(bgRgb24("Hello, World!", {
 *   r: 41,
 *   g: 42,
 *   b: 43,
 * }));
 * ```
 *
 * @module
 */

// deno-lint-ignore no-explicit-any
const { Deno } = globalThis as any;
const noColor = typeof Deno?.noColor === "boolean"
  ? Deno.noColor as boolean
  : true;
...
```

Maybe there is a way to get a JSON representation too! Not sure, but I think I remember that.

How can I correctly accomplish this same thing in Next.js? Is there a way to say the following?

> If header `Accept: application/json` is present, then call the `./api/[thing].js` to get the JSON response, whereas if the header is not present, return the rendered `./pages/[thing].jsx`.

Another way to do it would be to have two separate URLs, your `/api/[thing]` URL which returns JSON, and your `/[thing]` page URL for rendering HTML, but I'm wondering if we can accomplish something similar to what Deno.js is doing in some clever way.

If it's possible, that's great! In that case, what would you need to configure? If it's not possible, no problem, it would be good to know this is a limitation I can't have so I can stop thinking about it :)

Thanks!

**Comment by icyJoseph** [comment, marked as answer]
You can use rewrites: https://nextjs.org/docs/api-reference/next.config.js/rewrites#header-cookie-and-query-matching

And even rewrite at middleware level if you wish, and it'd also make it easier to prevent direct access to the JSON serving endpoint:

```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  async rewrites() {
    return {
      beforeFiles: [
        {
          source: "/",
          destination: "/api/hello",
          has: [
            { type: "header", key: "content-type", value: "application/json" },
          ],
        },
      ],
    };
  },
};

module.exports = nextConfig;
```

So if I had `data.json` with:

```json
{
  "foo": ["bar", "baz"]
}
```

And my landing page:

```tsx
import Head from "next/head";
import styles from "@/styles/Home.module.css";

import data from "../data.json";

export default function Home() {
  return (
    <>
      <Head>
        <title>Create Next App</title>
      </Head>
      <main className={styles.main}>
        <p>{data.foo}</p>
      </main>
    </>
  );
}
```

And my API route:

```ts
import type { NextApiRequest, NextApiResponse } from "next";

import data from "../../data.json";

export default function handler(req: NextApiRequest, res: NextApiResponse) {
  res.status(200).json(data);
}
```

Then when I request these, using [VsCode Rest Client](https://marketplace.visualstudio.com/items?itemName=humao.rest-client):

```http
###
GET http://localhost:3000

###
GET http://localhost:3000
Content-Type: application/json
```

I see HTML for the first one, and JSON for the second.
